File: codigo/haskell/spoj/EPALIN.hs
-- TODO: está todo feo el código

import           Data.Array (Array, elems, listArray, (!))

main :: IO ()
main = interact $ unlines . map proccess . lines

proccess xs =
    let rev = reverse xs
        aux = rev ++ '$':xs
        pi_array = map (\(_, x, _) -> x) (table aux)
        n = length xs -- igual y no uso este
        k = last pi_array
        other_part = drop k rev
    in xs ++ other_part


type Table a = [(a, Int, Int)]

table :: Eq a => [a] -> Table a
table as =
  [(a, p, n) | ((a, p), n) <- zip (elems (ptable as)) ([2 .. length as] ++ [0])]

ptable :: Eq a => [a] -> Array Int (a, Int)
ptable as = xa
  where
    xa = listArray (1, length as) (zip as (scanl op 0 (tail as)))
    op k b
      | a == b = k + 1
      | k == 0 = 0
      | otherwise = op p b
      where
        a = fst (xa ! (k + 1))
        p = snd (xa ! k)

File: codigo/cpp/spoj/EPALIN.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

vector<int> compute_prefix_function(const string &pattern) {
  int m = pattern.length();
  vector<int> pi(m);
  pi[0] = 0;
  int k = 0;

  for (int q = 1; q < m; q++) {
    while (k > 0 && pattern[k] != pattern[q])
      k = pi[k - 1];

    if (pattern[k] == pattern[q])
      k++;

    pi[q] = k;
  }
  return pi;
}

int main() {
  string s;
  while (cin >> s) {
    string ans = s;

    string rev = s;
    reverse(rev.begin(), rev.end());

    string aux = rev + '$' + s;

    vector<int> pi = compute_prefix_function(aux);
    int k = pi[aux.size() - 1];

    for (size_t i = k; i < s.length(); i++)
      ans += aux[i];

    cout << ans << endl;
  }
  return 0;
}
